src/advertisement/ticket.rs

```diff
@@ -0,0 +1,414 @@
+const MAX_RESPONSES_PER_REGTOPIC: u8 = 3;
+
+/// A topic is active when it associated with the node id from a node it is
+/// published on.
+#[derive(PartialEq, Eq, Hash, Clone)]
+pub struct ActiveTopic {
+    /// NodeId of the sender of the TICKET response.
+    node_id: NodeId,
+    /// The topic hash as it is sent in the TICKET response
+    topic: TopicHash,
+}
+
+impl ActiveTopic {
+    pub fn new(node_id: NodeId, topic: TopicHash) -> Self {
+        ActiveTopic { node_id, topic }
+    }
+
+    pub fn topic(&self) -> TopicHash {
+        self.topic
+    }
+
+    pub fn node_id(&self) -> &NodeId {
+        &self.node_id
+    }
+}
+
+/// A ticket is active when it is associated with the node contact of
+/// the sender of the ticket.
+pub struct ActiveTicket {
+    /// Node Contact of the sender of the ticket.
+    contact: NodeContact,
+    /// The ticket, an opaque object to the receiver.
+    ticket: Vec<u8>,
+}
+
+impl ActiveTicket {
+    pub fn new(contact: NodeContact, ticket: Vec<u8>) -> Self {
+        ActiveTicket { contact, ticket }
+    }
+
+    pub fn contact(&self) -> NodeContact {
+        self.contact.clone()
+    }
+
+    pub fn ticket(&self) -> Vec<u8> {
+        self.ticket.clone()
+    }
+}
+
+/// Tickets holds the tickets recieved in TICKET responses to locally
+/// initiated REGTOPIC requests.
+pub struct Tickets {
+    /// Tickets maps one ActiveTicket per ActiveTopic.
+    tickets: HashMapDelay<ActiveTopic, ActiveTicket>,
+    /// TicketHistory sets a time limit to how many times the ActiveTicket
+    /// value in tickets can be updated within a given ticket_limiter_duration.
+    ticket_history: TicketHistory,
+}
+
+impl Tickets {
+    pub fn new(ticket_limiter_duration: Duration) -> Self {
+        Tickets {
+            tickets: HashMapDelay::new(Duration::default()),
+            ticket_history: TicketHistory::new(ticket_limiter_duration),
+        }
+    }
+
+    pub fn insert(
+        &mut self,
+        contact: NodeContact,
+        ticket: Vec<u8>,
+        wait_time: Duration,
+        topic: TopicHash,
+    ) -> Result<(), &str> {
+        let active_topic = ActiveTopic::new(contact.node_id(), topic);
+
+        if let Err(e) = self.ticket_history.insert(active_topic.clone()) {
+            return Err(e);
+        }
+        self.tickets
+            .insert_at(active_topic, ActiveTicket::new(contact, ticket), wait_time);
+        Ok(())
+    }
+}
+
+impl Stream for Tickets {
+    type Item = Result<(ActiveTopic, ActiveTicket), String>;
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        match self.tickets.poll_next_unpin(cx) {
+            Poll::Ready(Some(Ok((active_topic, ticket)))) => {
+                Poll::Ready(Some(Ok((active_topic, ticket))))
+            }
+            Poll::Ready(Some(Err(e))) => {
+                debug!("{}", e);
+                Poll::Pending
+            }
+            Poll::Ready(None) => Poll::Pending,
```

Comment:
Awesome, thanks! Yes I got the idea to do this from the futures-challenge you gave me, where a stream called in a select! statement handled a list of values. When the list was empty/Poll::Ready(None) was returned, we returned Poll::Pending as it would fill up again.